### ISSUE TITLE
stream-line-column missing?

### ISSUE THREAD

**Issue by rtoy** [trivial-gray-streams/trivial-gray-streams#12]
As mentioned in http://www.nhplace.com/kent/CL/Issues/stream-definition-by-user.html, `stream-line-column` must be defined for every character output stream.  Perhaps trivial-gray-streams should include a defgeneric for it?  There might be other methods that are required; I didn't check.  But `stream-start-line-p` says you need to define that or `stream-fresh-line` if `stream-line-column` returns `nil`.  Don't know how that should be handled.

**Comment by avodonosov** [comment]
**Question 1**: Who should provide this method, the Gray Streams user code (e.g. the YASON library) or Gray Streams subsystem itself? 

Reading the Gray Streams proposal text, the wording at first sight makes impression it's the user code responsibility. It is surprising to find this. But for many other cases where default method implementation must be provided by the Gray Streams subsystem the doc  explicitly says "The default method ...".

The `stream-line-column` description does not have the phrase "The default method."

On the other hand, the description says
>  For every character output stream class that is defined,
  a method must be defined for this function, although it is
  permissible for it to always return NIL.

The `fundamental-character-stream` class is (pre-)defined by the Gray Streams subsystem. Does this requirement of having a method apply to apply `fundamental-character-stream`? If it does, then the Gray Streams subsystem must provide the method.

Later in the Discussion section of the doc, we read
> There are some differences of opinion about whether
  certain operations should have default methods or required methods:
  STREAM-LISTEN, STREAM-READ-CHAR-NO-HANG, STREAM-LINE-COLUMN,
  and STREAM-START-LINE-P.

**Question 2**: If that's a responsibility of the Gray Streams subsystem, does it mean trivial-gray-streams or the underlying Gray Streams in a lisp implementation like CMUCL?

Currently trivial-gray-streams only implements methods only for `stream-position` and `stream-read-sequence` / `stream-write-sequence` - these methods absent in the Gray proposal (most likely because their counterparts in the common-lisp package are absent in the "Common Lisp the Lanuguage" book).  So the lisp implementations differ in generic function names and parameters for such functions, and trivial-gray-streams defines a portable interface for them, creates methods for implementation-specific functions from which it  calls  its  portable version.

In a similar fashion,  since we see in practice the lisp implementations differ in existence of default method for `stream-line-column`, trival-gray-streams could be a point of unification. It can define a default method for `(stream-line-column fundamental-character-stream)` to return NIL on all implementations, or to signal an "user code must define stream-line-column" error on all implementations.

I personally see only advantages in having NIL default method. What reasons could have those people mentioned by the Discussion section of the Gray proposal, who were against the default method for `stream-line-column'?

**Comment by rtoy** [comment]
I'm not really sure; I've never really used gray streams before.  

However, it seems useful for trivial-gray-streams to re-export `stream-line-column` from the implementation-specific package so that user code doesn't need to know about the implementation package. I guess this goes for all the other methods in gray-streams?

**Comment by avodonosov** [comment]
Yes, `stream-line-column` and alll other gray function symbols are re-exported.

**Comment by rtoy** [comment]
Oh, sorry, I didn't notice that it was re-exported.  So we're all good here.  I don't think trivial-gray-streams needs to define anything.  It's up to the user to define a method for `stream-line-column`.

Closing issue.  Thanks for your help!

**Comment by avodonosov** [comment]
As a portability layer, trivial-gray-steams could shield users from difference in treatments of Gray proposals by lisp implementations. To avoid cases when a code written and tested on an implementation where the default NIL method exists, suddenly fails on another implementation, as this happened with the YASON library. 

I will reopen the issue, at least to keep this question move visible for now.

**Comment by avodonosov** [comment]
BTW, regarding this quote from the Gray proposal:

> There are some differences of opinion about whether
certain operations should have default methods or required methods:
STREAM-LISTEN, STREAM-READ-CHAR-NO-HANG, STREAM-LINE-COLUMN,
and STREAM-START-LINE-P.

and my question "What reasons could have those people mentioned in the Discussion section of the Gray proposal, who were against the default method for stream-line-column"

I guess for STREAM-LISTEN, STREAM-READ-CHAR-NO-HANG the reason is that the default methods are blocking (because implemented in terms of STREAM-READ-CHAR). The reasons for STREAM-START-LINE-P and STREAM-LINE-COLUMN are still unclear to me.

Of course, one thing is to not have default methods of these functions, for whatever reasons, and another thing is to sometimes (on some lisp implementations) have them, and sometimes not (on other lisp implementations). Such unpredictable situations is probably worse than clear choice for presence or absence of the default NIL method.

It is interesting that of the four functions only STREAM-LINE-COLUMN description is so vague. For the other three, the doc says exactly what default method does.

**Comment by rtoy** [comment]
For cmucl, I see that it's implementation defines methods for `stream-start-line-p` and `stream-advance-to-column` for `fundamental-character-output-streams`.  These methods call `stream-line-column`.

I think cmucl should provide a method for `stream-line-column` for a `fundamental-character-output-stream`.  See [cmucl issue 147](https://gitlab.common-lisp.net/cmucl/cmucl/-/issues/147)
